ARROW-17407: [Doc][FlightRPC] Flight/gRPC best practices

rok opened the pull request:
We want to provide best practices and debugging section in:

[cpp flight docs](https://arrow.apache.org/docs/cpp/flight.html)
[python flight docs](https://arrow.apache.org/docs/python/flight.html)
[java flight docs](https://arrow.apache.org/docs/java/flight.html)
[R flight docs](https://arrow.apache.org/docs/r/articles/flight.html)

zeroshade commented:
@rok  Can we also add best practices and debugging sections in the Go docs too?

This can be done by either a README.md in the `go/arrow/flight` directory or by having a "package comment" in the directory (a comment directly above the `package flight` line in one of the .go files (convention is usually a file named `doc.go`).

Would you prefer for me to add it to this branch?

rok commented:
> @rok Can we also add best practices and debugging sections in the Go docs too?

Thanks for chiming in @zeroshade! I was going to ask you where to put the Go specific content :).
I'll add it and ping you back.

zeroshade commented:
@rok Sounds great! Have a look at `go/arrow/doc.go` or `go/parquet/doc.go` for some examples.

https://tip.golang.org/doc/comment also gives a good rundown of how the format of the comment gets translated into the formatting of the rendered docs on pkg.go.dev (TL;DR it's similar to just using markdown)

rok commented:
@zeroshade I've added the Go doc. Please check if it makes sense, it's my first time touching Go docs.

rok commented:
@lidavidm I'm now linking to c++ from python/java/R docs which is not ideal, but duplicating text is not either. Any thoughts?

drin commented:
> I'm now linking to c++ from python/java/R docs which is not ideal, but duplicating text is not either. Any thoughts?

IMO, I would think there should be a common area for the text since that is common across languages. Then, the c++ docs can include code snippets as appropriate which don't seem too relevant to the other languages.

The "[Arrow Flight RPC](https://arrow.apache.org/docs/format/Flight.html)" section under "Specifications and Protocols" seems like a good place to me, since those descriptions seem to be more related to best practices of using the protocol overall.

drin commented:
The code snippet could potentially be moved to a cookbook type thing too?

lidavidm commented:
What's wrong with linking to the C++ docs?

The format/ page for Flight RPC is meant more for things all implementations agree on, I don't think these tips really go there

lidavidm commented:
Also, linking to the C++ docs isn't applicable in all situations, they don't apply to Java for instance.

lidavidm commented:
The same goes for Go.

If we want to centralize things, we should think about where to do that. And then all the language docs can link there. (I think Sphinx has a 'tabs' extension we could use for the language-specific snippets too.)

drin commented:
> What's wrong with linking to the C++ docs?

I don't think there's anything wrong with it. I was just thinking through alternatives since Rok didn't seem to like it.

> Also, linking to the C++ docs isn't applicable in all situations, they don't apply to Java for instance.

Maybe I misread that portion of the PR, but I think it references it at the moment? I don't know the details, so not sure if some of the descriptions need disentangling

lidavidm commented:
The Java docs link to the C++ docs - but the C++ docs don't apply to Java. The bit about allocations doesn't apply to Java at all, for instance.

drin commented:
options that make the most sense to me are:
1. make a new page, perhaps under the "development" section?
2. leave the structure as is, and let a PR for ARROW-17378 and/or ARROW-6390 re-structure?

lidavidm commented:
Maybe we should consider a separate tutorial section then? Even if they're language-specific? (Well, 'tutorials and guides' or something…) The Flight tutorials could probably be made to work with the same 'tab' structure so they can cover multiple languages at once (though that should be used sparingly)

drin commented:
Assuming [this tab structure](https://sphinx-tabs.readthedocs.io/en/latest/):
I like the idea of a tab structure in cases where code snippets are small, which seems to apply to what I see in this documentation. I also think that for something like flight, the focus is more on the protocol and communication so it's a case where there is a lot of conceptual overlap so tabs may work well here.

Otherwise, I don't have strong opinions about "best practices" being in, or separate from, a tutorial section.

lidavidm commented:
I didn't get to this today, should hopefully get to look at it this week

wjones127 commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,110 @@ request/response. On the server, they can inspect incoming headers and
+bounded thread pool -> Reject connections / requests when under load, and have
+clients retry with backoff. This also gives an opportunity to retry with a
+different node. Not everyone gets serviced but quality of service stays consistent.
+
```
```suggestion

Limiting DoPut Batch Size
--------------------------

You may wish to limit the maximum size a client can submit to a server through
DoPut, to prevent a request from taking up too much memory on the server. On
the client-side, set `write_size_limit_bytes` on `FlightClientOptions`. On the 
server-side, set the gRPC option `GRPC_ARG_MAX_RECEIVE_MESSAGE_LENGTH`.
The client-side option will return an error that can be retried with smaller batches,
while the server-side limit will close out the connection. Setting both can be wise, since
the former provides a better user experience but the latter may be necessary to defend
against impolite clients.

```

wjones127 reviewed (commented):
Thank you for starting this @rok! I've added a few suggestions.

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,110 @@ request/response. On the server, they can inspect incoming headers and
+gRPC
+----
+
+When using default gRPC transport options can be passed to it via
+:member:`arrow::flight::FlightClientOptions::generic_options`. For example:
```
This is less a best practice, more of a "howto" so I think this should go in the cookbook

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,110 @@ request/response. On the server, they can inspect incoming headers and
+Don’t round-robin load balance
+------------------------------
+
+Round robin balancing can cause every client to have an open connection to
```
Link to the gRPC page describing load balancing? https://github.com/grpc/grpc/blob/master/doc/load-balancing.md#round_robin

rok replied:
Added.

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,110 @@ request/response. On the server, they can inspect incoming headers and
+Don’t round-robin load balance
+------------------------------
+
+Round robin balancing can cause every client to have an open connection to
+every server causing an unexpected number of open connections and a depletion
+of resources.
```
```suggestion
Round robin balancing means every client can have an open connection to
every server, causing an unexpected number of open connections and depleting
server resources.
```

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,110 @@ request/response. On the server, they can inspect incoming headers and
+headers (on both sides) so you can see if grpc established the connection or
+not. It will also print when a message is sent, so you can tell if the
+connection is open or not.
+Note: "connect" isn't really a connect and we’ve observed that gRPC does not
+give you the actual error until you first try to make a call. This can cause
+error being reported at unexpected times.
```
```suggestion
gRPC may not report connection errors until a call is actually made.
Hence, to detect connection errors when creating a client, some sort
of dummy RPC should be made.
```

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,110 @@ request/response. On the server, they can inspect incoming headers and
+give you the actual error until you first try to make a call. This can cause
+error being reported at unexpected times.
+
+Use ListFlights sparingly
```
This is too specific to have, IMO

rok replied:
Removed from here and from Go doc.

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,108 @@ request/response. On the server, they can inspect incoming headers and
+Memory cache client-side
+------------------------
+
+Flight uses gRPC allocator wherever possible.
```
```suggestion
Flight tries to reuse allocations made by gRPC to avoid redundant
data copies.  However, this means that those allocations may not
be tracked by the Arrow memory pool, and that memory usage behavior,
such as whether free memory is returned to the system, is dependent
on the allocator that gRPC uses (usually the system allocator).
```

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,110 @@ request/response. On the server, they can inspect incoming headers and
+caches in some workloads. But you can explicitly tell malloc to dump caches.
+See ARROW-16697_ as an example.
+
+A quick way of testing: attach to the process with a debugger and call malloc_trim
```
or call ReleaseUnused on the system pool

rok replied:
Added.

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,110 @@ request/response. On the server, they can inspect incoming headers and
+yet implemented.
+
+
+Memory cache client-side
```
this isn't specific to the client

rok replied:
Changed.

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,110 @@ request/response. On the server, they can inspect incoming headers and
+
+Flight uses gRPC allocator wherever possible.
+
+gRPC will spawn an unbounded number of threads for concurrent clients. Those
```
Move this to 'excessive traffic' below?

rok replied:
Done I think. Diff is confusing me right now.

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,110 @@ request/response. On the server, they can inspect incoming headers and
+This is what you are seeing now.
+bounded thread pool -> Reject connections / requests when under load, and have
+clients retry with backoff. This also gives an opportunity to retry with a
+different node. Not everyone gets serviced but quality of service stays consistent.
```
we should add code to configure these modes

rok replied:
You mean like setting `GRPC_ARG_MIN_RECONNECT_BACKOFF_MS` to `200ms` for bounded and `0ms` for unbounded? Perhaps another setting I'm not aware of?

rok replied:
Added `GRPC_ARG_MIN_RECONNECT_BACKOFF_MS` example.

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,110 @@ request/response. On the server, they can inspect incoming headers and
+
+1. A stale connection can be closed using
+   :member:`arrow::flight::FlightClientOptions::stop_token`. This requires recording the
+   stop token at connection establishment time.
```
this needs code samples

(it would be great if all code samples linked to the cookbook)

rok replied:
Perhaps we could add samples for now and they would be moved out as cookbooks become available?
Also I don't see an issue if we duplicate atomic pieces of the cookbooks here.

rok replied:
Yeah, agreed.

lidavidm replied:
We can maybe merge this then add issues for moving each of these examples to the cookbook.

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,110 @@ request/response. On the server, they can inspect incoming headers and
 
+.. _flight-best-practices:
+
+Best practices
```
Link to the grpc best practices doc as well?

rok replied:
They are linked below as `best gRPC practices`

zeroshade commented on go/arrow/flight/doc.go:
```diff
@@ -0,0 +1,98 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Here we list best practices and common pitfalls for Arrow Flight usage.
+//
+// GRPC
```
Typically the package comment should begin with `Package <name>` so we should first start with:

`Package flight contains server and client implementations for the Arrow Flight RPC` and any other description of the package. Then follow to the best practices and pitfalls after that.

The reason for this is that the first sentence will show up in the documentation as a summary for the package:

![image](https://user-images.githubusercontent.com/555095/185425569-f268f22a-f66c-4463-9737-72ccf0f3da03.png)

rok replied:
Got it! That's nice to know. Writing go docs for the first time is interesting.

zeroshade commented on go/arrow/flight/doc.go:
```diff
@@ -0,0 +1,98 @@
+// http2 logs (using 2 is more verbose with frame dumps). This will print the
+// initial headers (on both sides) so you can see if grpc established the
+// connection or not. It will also print when a message is sent, so you can tell
+// if the connection is open or not.
+// Note: "connect" isn't really a connect and we’ve observed that gRPC does not
```
can you put a blank space between these paragraphs please?

zeroshade commented on go/arrow/flight/doc.go:
```diff
@@ -0,0 +1,98 @@
+//
+// ListFlights endpoint is largely just implemented as a normal GRPC stream
+// endpoint and can hit transfer bottlenecks if used too much. To estimate data
+// transfer bottleneck:
+// 5k schemas will serialize to about 1-5 MB/call. Assuming a gRPC localhost
+// bottleneck of 3GB/s you can at best serve 600-3000 clients/s.
```
```suggestion
// transfer bottleneck:
//
// 5k schemas will serialize to about 1-5 MB/call. Assuming a gRPC localhost
// bottleneck of 3GB/s you can at best serve 600-3000 clients/s.

zeroshade commented on go/arrow/flight/doc.go:
```diff
@@ -0,0 +1,98 @@
+// 5k schemas will serialize to about 1-5 MB/call. Assuming a gRPC localhost
+// bottleneck of 3GB/s you can at best serve 600-3000 clients/s.
+//
+// https://issues.apache.org/jira/browse/ARROW-15764 proposes a caching
+// optimisation for server side, but it was not yet implemented.
```
probably should mention that the caching optimization is specifically for C++ in that issue.

zeroshade commented on go/arrow/flight/doc.go:
```diff
@@ -0,0 +1,98 @@
+//
+// Flight uses gRPC allocator wherever possible.
+//
+// gRPC will spawn an unbounded number of threads for concurrent clients. Those
+// threads are not necessarily cleaned up (cached thread pool in java parlance).
+// glibc malloc clears some per thread state and the default tuning never clears
+// caches in some workloads. But you can explicitly tell malloc to dump caches.
+// See https://issues.apache.org/jira/browse/ARROW-16697 as an example.
```
This would actually be incorrect for the Go implementation since it uses goroutines which do not have a 1-to-1 relationship with OS threads. It wouldn't fit the "thread pool" description. We can probably remove this entire paragraph for the Go docs here.

rok replied:
Changed.

zeroshade reviewed (changes requested):
Looks pretty good to me on the Go side though some of the comments aren't accurate or relevant to the Go library, i've made some suggestions.

rok commented on go/arrow/flight/doc.go:
```diff
@@ -0,0 +1,98 @@
+// (easily) possible to implement with the blocking gRPC API. For now one can also do
+// something like set up a background thread that calls cancel() on a timer and have
+// the main thread reset the timer every time a write operation completes successfully
+// (that means one needs to use to_batches() + write_batch and not write_table).
```
Added something, but I'm not sure I understand enough for it to make sense :).

rok reviewed (commented):
Thanks for the review everyone! I've gone through your comments and hopefully used them productively.

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,154 @@ request/response. On the server, they can inspect incoming headers and
+         .. code-block:: python
+
+
+2. Use client timeout.
```
```suggestion
2. Use call timeouts. (This is a general gRPC best practice.)
```

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,154 @@ request/response. On the server, they can inspect incoming headers and
+
+         .. code-block:: python
+
+3. There is a long standing ticket for a per-write/per-read timeout instead of a per
```
This first sentence might be better as a `.. note` since it's not actually actionable

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,154 @@ request/response. On the server, they can inspect incoming headers and
+
+3. There is a long standing ticket for a per-write/per-read timeout instead of a per
+   call timeout (ARROW-6062_), but this is not (easily) possible to implement with the
+   blocking gRPC API. For now one can also do something like set up a background thread
```
This sort of thing should definitely be a recipe.

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,154 @@ request/response. On the server, they can inspect incoming headers and
+
+         .. code-block:: python
+
+3. There is a long standing ticket for a per-write/per-read timeout instead of a per
+   call timeout (ARROW-6062_), but this is not (easily) possible to implement with the
+   blocking gRPC API. For now one can also do something like set up a background thread
+   that calls cancel() on a timer and have the main thread reset the timer every time a
+   write operation completes successfully (that means one needs to use to_batches() +
+   write_batch and not write_table).
```
```suggestion
3. Client timeouts are not great for long-running streaming calls, where it may
   be hard to choose a timeout for the entire operation. Instead, what is often
   desired is a per-read or per-write timeout so that the operation fails if it
   isn't making progress. This can be implemented with a background thread that
   calls cancel() on a timer, with the main thread resetting the timer every time
   an operation completes successfully. For a fully-worked out example, see the
   Cookbook.
   
   .. note:: There is a long standing ticket for a per-write/per-read timeout
             instead of a per call timeout (ARROW-6062_), but this is not (easily)
             possible to implement with the blocking gRPC API. 
```

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,154 @@ request/response. On the server, they can inspect incoming headers and
+Hence, to detect connection errors when creating a client, some sort
+of dummy RPC should be made.
+
+Memory cache management
```
```suggestion
Memory management
```

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,154 @@ request/response. On the server, they can inspect incoming headers and
+such as whether free memory is returned to the system, is dependent
+on the allocator that gRPC uses (usually the system allocator).
+
+A quick way of testing: attach to the process with a debugger and call malloc_trim
+or call ReleaseUnused on the system pool.
```
```suggestion
A quick way of testing: attach to the process with a debugger and call
malloc_trim, or call ReleaseUnused on the system pool. If memory usage
drops, then likely, there is memory allocated by gRPC or by the
application that the system allocator was holding on to. This can be
adjusted in platform-specific ways; see an investigation in JIRA for
an example of how this works on Linux/glibc. 
```

and link the JIRA

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,154 @@ request/response. On the server, they can inspect incoming headers and
+Excessive traffic
+-----------------
+
+gRPC will spawn an unbounded number of threads for concurrent clients. Those
```
This might not actually be true? e.g. `https://github.com/grpc/grpc/issues/14492`

Ah, it's not actually _unbounded_, just a very large number: https://github.com/grpc/proposal/blob/master/L30-cpp-control-max-threads-in-SyncServer.md

I think the threads are cleaned up now? So the next sentence may not be right

rok replied:
Updated. Please see if it makes sense.

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,154 @@ request/response. On the server, they can inspect incoming headers and
+
+gRPC will spawn an unbounded number of threads for concurrent clients. Those
+threads are not necessarily cleaned up (cached thread pool in java parlance).
+glibc malloc clears some per thread state and the default tuning never clears
+caches in some workloads. But you can explicitly tell malloc to dump caches.
+See ARROW-16697_ as an example.
```
This might belong better in the discussion of memory management below

rok replied:
Moved.

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,154 @@ request/response. On the server, they can inspect incoming headers and
+caches in some workloads. But you can explicitly tell malloc to dump caches.
+See ARROW-16697_ as an example.
+
+There are basically two ways to handle excessive traffic:
+* unbounded thread pool -> everyone gets serviced, but it might take forever.
+* bounded thread pool -> Reject connections / requests when under load, and have
+clients retry with backoff. This also gives an opportunity to retry with a
+different node. Not everyone gets serviced but quality of service stays consistent.
```
I'd probably word this like ~gRPC's default behavior allows one server to accept many connections from many different clients, but if requests do a lot of work (as they may under Flight), the server may not be able to keep up. Configuring the server to limit the number of clients and reject requests more proactively, and configuring clients to retry with backoff (and potentially connect to a different node), would give more consistent quality of service.

rok replied:
Switched to proposed text.

zeroshade commented on go/arrow/flight/doc.go:
```diff
@@ -0,0 +1,79 @@
+//
+// Excessive traffic
+//
+// There are basically two ways to handle excessive traffic:
+// * unbounded thread pool -> everyone gets serviced, but it might take forever.
+// This is what you are seeing now.
+// bounded thread pool -> Reject connections / requests when under load, and have
+// clients retry with backoff. This also gives an opportunity to retry with a
+// different node. Not everyone gets serviced but quality of service stays consistent.
```
Let's change this to `unbounded goroutines` (where you get limited by `GOMAXPROCS` i.e.: number of cores) vs a limited set of workers that handle streams. You can control this with https://pkg.go.dev/google.golang.org/grpc#NumStreamWorkers where the default (0) will simply spawn a new goroutine for every request stream.

rok replied:
Got it! Changed.

zeroshade commented on go/arrow/flight/doc.go:
```diff
@@ -0,0 +1,79 @@
+//
+// Closing unresponsive connections
+//
+// * A stale connection can be closed using CallOptions.stop_token. This requires
+// recording the stop token at connection establishment time.
```
As far as I can tell, there doesn't appear to be a `stop_token` for the Go implementation, stale connection cleanup is handled in the background by the `Client`

rok replied:
Removed.

zeroshade commented on go/arrow/flight/doc.go:
```diff
@@ -0,0 +1,79 @@
+//
+// * A stale connection can be closed using CallOptions.stop_token. This requires
+// recording the stop token at connection establishment time.
+// * Use client timeout
```
Replace this with the comment about `context.Context`, there's no separate "client timeout" the intent is to leverage the existing Go context API

rok replied:
Done.

zeroshade commented on go/arrow/flight/doc.go:
```diff
@@ -0,0 +1,79 @@
+// something like set up a background thread that calls cancel() on a timer and have
+// the main thread reset the timer every time a write operation completes successfully
+// (that means one needs to use to_batches() + write_batch and not write_table).
+// * context.Context can be passed to requests to provide (https://pkg.go.dev/context#WithTimeout) or
+// (https://pkg.go.dev/context#WithCancel).
```
the `context.Context` object is required, not optional. If a user doesn't have a context, they should pass `context.Background()`, they should never pass a nil context (as per the Go docs)

rok replied:
Got it! Changed. Please see if it makes sense.

rok commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,154 @@ request/response. On the server, they can inspect incoming headers and
+Re-use clients whenever possible
+--------------------------------
+
+Closing clients causes gRPC to close and clean up connections which can take
```
I'm not sure what's more costly time wise - creating a client or closing it. I don't know that somebody measured this. Either way it's remedied by re-using clients.

lidavidm replied:
Maybe word it more generically, ~"Creating and closing clients requires setup/teardown on the client and server which can take away from actually handling RPCs. Reuse clients whenever possible to avoid this. Note that clients are thread-safe."

rok commented:
@github-actions crossbow submit preview-docs

rok commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,154 @@ request/response. On the server, they can inspect incoming headers and
+Debugging
+---------
+
+Use netstat to see the number of open connections.
```
Added.

rok commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,154 @@ request/response. On the server, they can inspect incoming headers and
+
+You may wish to limit the maximum size a client can submit to a server through
+DoPut, to prevent a request from taking up too much memory on the server. On
+the client-side, set `write_size_limit_bytes` on `FlightClientOptions`. On the 
```
Added.

rok commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,154 @@ request/response. On the server, they can inspect incoming headers and
+
+      .. code-block:: cpp
+
+         auto options = FlightClientOptions::Defaults();
```
Changed. Please check if it makes sense.

rok commented:
https://crossbow.voltrondata.com/pr_docs/13873/

rok commented:
@github-actions crossbow submit preview-docs

rok commented:
@lidavidm @zeroshade I've addressed most of the feedback, you can see the rendered docs here: https://crossbow.voltrondata.com/pr_docs/13873/cpp/flight.html#best-practices

I'll be traveling until next Monday so feel free to either wait or finish this without me.

rok commented:
@github-actions crossbow submit preview-docs

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,198 @@ request/response. On the server, they can inspect incoming headers and
+on to. This can be adjusted in platform-specific ways; see an investigation
+in JIRA for an example of how this works on Linux/glibc.
+
+malloc can be explicitly told to dump caches. See ARROW-16697_ as an example.
```
glibc malloc?

rok commented:
@github-actions crossbow submit preview-docs

rok commented:
I've addressed the feedback (hopefully all) and opened a cookbook issue (https://github.com/apache/arrow-cookbook/issues/251).
Would appreciate another look! :)

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,198 @@ request/response. On the server, they can inspect incoming headers and
+every server, causing an unexpected number of open connections and depleting
+server resources.
+
+Debugging disconnects
```
Maybe just change this to "Debugging connection issues"?

rok replied:
Done.

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,198 @@ request/response. On the server, they can inspect incoming headers and
+on the system pool. If memory usage drops, then likely, there is memory
+allocated by gRPC or by the application that the system allocator was holding
+on to. This can be adjusted in platform-specific ways; see an investigation
+in JIRA for an example of how this works on Linux/glibc.
```
Which jira?

rok replied:
We were referring to the same on twice I think. Changed.

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,198 @@ request/response. On the server, they can inspect incoming headers and
+on the allocator that gRPC uses (usually the system allocator).
+
+A quick way of testing: attach to the process with a debugger and call
+malloc_trim, or call :func:`ReleaseUnused <arrow::MemoryPool::ReleaseUnused>`
```
```suggestion
``malloc_trim``, or call :func:`ReleaseUnused <arrow::MemoryPool::ReleaseUnused>`
```

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,198 @@ request/response. On the server, they can inspect incoming headers and
+gRPC's default behavior allows one server to accept many connections from many
+different clients, but if requests do a lot of work (as they may under Flight),
+the server may not be able to keep up. Configuring the server to limit the number
+of clients and reject requests more proactively, and configuring clients to retry
```
How do you configure the server to do this?

lidavidm replied:
(IIRC, it's not possible without modifying Flight…)

rok replied:
I think this is referring to gRPC settings.

lidavidm replied:
I would say either remove this text or have a code sample of how to set the server-side limit

rok replied:
[See here](https://github.com/apache/arrow/pull/13873#discussion_r972262181)

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,198 @@ request/response. On the server, they can inspect incoming headers and
+Limiting DoPut Batch Size
+--------------------------
+
+You may wish to limit the maximum size a client can submit to a server through
```
```suggestion
You may wish to limit the maximum batch size a client can submit to a server through
```

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,198 @@ request/response. On the server, they can inspect incoming headers and
+Closing unresponsive connections
+--------------------------------
+
+1. A stale connection can be closed using
```
This is more a way to stop a long-running RPC call, than a connection specifically

lidavidm replied:
Actually all of these are so the section should just be renamed, and remove/reword any text referring to connections

rok replied:
Done.

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,198 @@ request/response. On the server, they can inspect incoming headers and
+--------------------------------
+
+1. A stale connection can be closed using
+   :member:`arrow::flight::FlightClientOptions::stop_token`. This requires recording the
```
This is FlightCallOptions not ClientOptions

rok replied:
I think this is already `FlightCallOptions` just the diff is off.

lidavidm replied:
I still see ClientOptions here

rok replied:
Saw it here now too. Fixed.

rok commented:
Thanks for the review @lidavidm ! I think I addressed everything and sorry for not being faster.

rok commented:
@github-actions crossbow submit preview-docs

rok commented:
https://crossbow.voltrondata.com/pr_docs/13873/cpp/flight.html#best-practices

lidavidm commented:
sorry, I'm fairly backed up, I'll try to get to this sometime this week

rok commented:
No hurry on my end @lidavidm! Sorry for the slow iterations.

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,197 @@ request/response. On the server, they can inspect incoming headers and
+gRPC
+----
+
+When using default gRPC transport options can be passed to it via
```
```suggestion
When using the default gRPC transport, options can be passed to it via
```

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,197 @@ request/response. On the server, they can inspect incoming headers and
+
+Creating and closing clients requires setup and teardown on the client and
+server side which can take away from actually handling RPCs. Reuse clients
+whenever possible to avoid this. Note that clients are thread-safe.
```
```suggestion
whenever possible to avoid this. Note that clients are thread-safe, so a
single client can be shared across multiple threads.
```

lidavidm commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,197 @@ request/response. On the server, they can inspect incoming headers and
+   be hard to choose a timeout for the entire operation. Instead, what is often
+   desired is a per-read or per-write timeout so that the operation fails if it
+   isn't making progress. This can be implemented with a background thread that
+   calls cancel() on a timer, with the main thread resetting the timer every time
```
```suggestion
   calls Cancel() on a timer, with the main thread resetting the timer every time
```

rok commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,198 @@ request/response. On the server, they can inspect incoming headers and
+--------------------------------
+
+1. A stale call can be closed using
+   :member:`arrow::flight::FlightClientOptions::stop_token`. This requires recording the
```
```suggestion
   :member:`arrow::flight::FlightCallOptions::stop_token`. This requires recording the
```

rok commented on docs/source/cpp/flight.rst:
```diff
@@ -172,6 +172,198 @@ request/response. On the server, they can inspect incoming headers and
+glibc malloc clears some per thread state and the default tuning never clears
+caches in some workloads.
+
+gRPC's default behavior allows one server to accept many connections from many
+different clients, but if requests do a lot of work (as they may under Flight),
+the server may not be able to keep up. Configuring the server to limit the number
+of clients and reject requests more proactively, and configuring clients to retry
```
How about this:
```suggestion
gRPC's default behaviour allows one server to accept many connections from many
different clients, but if requests do a lot of work (as they may under Flight),
the server may not be able to keep up. Configuring clients to retry
```

rok commented:
@lidavidm thanks for the review! I've merged you change proposals.

lidavidm reviewed (approved):
Thanks!

lidavidm commented:
I think in terms of followups:

- Can we configure a server to shed load more proactively?
- Porting examples to the cookbook
- Providing a recipe for a per-read timeout

rok commented:
@lidavidm shall I open cookbook github issues for those?

lidavidm commented:
I think you did already but if not, yeah

rok commented:
Indeed: https://github.com/apache/arrow-cookbook/issues/251
